### ISSUE TITLE
False positive no-semicolon

### ISSUE THREAD

**Issue by nicolasamaduro** [pinterest/ktlint#1933]
In Kotlin, if you want to add properties or functions to an enum class, you have to separate their constants with a `;` from the rest of the body ([ref](https://kotlincompact.com/no-semicolon.html)):

```kotlin
enum class Weekdays {
  MONDAY, TUESDAY, WEDNESDAY, THURSDAY, FRIDAY, SATURDAY, SUNDAY;

  fun isWeekend() = this == SATURDAY || this == SUNDAY
}
```

This semicolon is throwing a false positive [no-semi](https://pinterest.github.io/ktlint/rules/standard/#no-semicolons) rule and it also messes up with the [trailing-comma-on-declaration-site](https://pinterest.github.io/ktlint/rules/standard/#trailing-comma-on-declaration-site) rule.

**Comment by paul-dingemans** [comment]
The problem is not reproducible as too little information is given.
- What is the ktlint version?
- What are your `.editorconfig` settings?
- What is the output produced?

You might also want to check whether your problem is reproducible with current [snapshot versions](https://pinterest.github.io/ktlint/install/snapshot-build/) in which some problems with enum have been resolved already.

**Comment by hfhbd** [comment]
At least our problem was fixed with this kotlin code:
```kotlin
package app.cash.sqldelight.core.integration

enum class Shoots {
  LEFT,
  RIGHT, // added manually
  ;

  enum class Type {
    ONE,
    TWO,
  }
}
```
While I absolutely understand the new trailing comma, the error message is hard to debug.

**Comment by paul-dingemans** [comment]
> * What went wrong:
> Execution failed for task ':spotlessKotlin'.
> > Error on line: 1, column: 1
>   rule: trailing-comma-on-declaration-site
>   Missing trailing comma before "}"

The output above is produced by Spotless. The Ktlint CLI (lint only) produces output like below:
```
src/main/kotlin/Foo.kt:5:10: Missing trailing comma before ";" (trailing-comma-on-declaration-site)
src/main/kotlin/Foo.kt:9:12: Missing trailing comma before "}" (trailing-comma-on-declaration-site)
```
which contains both the line number as well as the offset on the line.

When running Ktlint CLI with formatting, the output below is produced:
```
src/main/kotlin/Foo.kt:1:1: Missing trailing comma before "}" (trailing-comma-on-declaration-site)
src/main/kotlin/Foo.kt:7:4: Unnecessary semicolon (no-semi)
```
which indeed is not correct and already has been fixed in the [snapshot](https://pinterest.github.io/ktlint/install/snapshot-build/) of the upcoming `0.49.x` release. `0.48.2` however did format the code correctly.

**Comment by stkent** [comment]
I ran into this same case, I think, and filed an issue on the Spotless repo. https://github.com/diffplug/spotless/issues/1699

_Should_ `ktlint -F` be exiting with a non-zero exit code when it successfully reformats source code? (I am not familiar enough with standalone ktlint to know the norms around its exit codes.) e.g. when I run `ktlint -F` (version 0.48.2) on this file:

```kotlin
enum class AccountType {
    Case1,
    Case2;

    companion object {
    }
}
```

is it expected that both (1) the code is successfully reformatted (see below) and (2) the exit code is 1?

```kotlin
enum class AccountType {
    Case1,
    Case2,
    ;

    companion object {
    }
}
```

Thanks in advance for any light you can shed on this!

**Comment by paul-dingemans** [comment]
In case all lint errors have been autocorrected, `ktlint -F` returns with exit code 0. In case at least one error could not be autocorrected, `ktlint -F` returns exit code 1. Note that `ktlint` (without `-F`) returns exit code 1 whenever a lint error is found regardless whether it can or cannot be autocorrected.

**Comment by stkent** [comment]
> ```kotlin
> AccountType
> ```

Thanks for the clarification for expected behavior, @paul-dingemans! Given this information, this sequence of events looks like it shows a bug?

* Initial source file:
    ```kotlin
    enum class AccountType {
        Case1,
        Case2,
        ;
    
        companion object {
        }
    }
    ```
* Run `ktlint -F path/to/AccountType.kt`.
    * Output:
        ```
        .../AccountType.kt:1:1: Missing trailing comma before "}" (trailing-comma-on-declaration-site)
        .../AccountType.kt:5:4: Unnecessary semicolon (no-semi)
        
        Summary error count (descending) by rule:
          no-semi: 1
          trailing-comma-on-declaration-site: 1
        ```
    * Exit code is `1`, implying that _at least one error could not be autocorrected_.
    * Updated source file:
        ```
        enum class AccountType {
            Case1,
            Case2,
            ;
        
            companion object {
            }
        }
        ```
    * Run `ktlint -F path/to/AccountType.kt` again. Since the previous invocation of ktlint indicated that _at least one error could not be autocorrected_, this should result in exit code `1`. But it does not; the exit code is **`0`**. This contradicts the assertion that a non-correctable error exists in the updated source file.

Perhaps it is the case that in fixing a correctable error, the error that could not (in isolation) be autocorrected is "accidentally" also fixed?

**Comment by paul-dingemans** [comment]
> Run `ktlint -F path/to/AccountType.kt`.
> * Output:
>   ```
>   .../AccountType.kt:1:1: Missing trailing comma before "}" (trailing-comma-on-declaration-site)
>   .../AccountType.kt:5:4: Unnecessary semicolon (no-semi)
>   
>   Summary error count (descending) by rule:
>     no-semi: 1
>     trailing-comma-on-declaration-site: 1
>   ```

Are you sure that you actually did run with `-F`? If so, I would not have expected to see the lint violation being reported. For me the command `ktlint -F path/to/AccountType.kt` returns `0`. Without `-F` I do see the output you mentioned and the return code is indeed `1`.

**Comment by stkent** [comment]
I believe that I did!

<img width="1048" alt="CleanShot 2023-05-11 at 15 15 53@2x" src="https://github.com/pinterest/ktlint/assets/6463980/d2153e96-194e-4677-8b54-e3e37d92e32a">

Note (in case this note was not obvious in my original comment) that this is with ktlint 0.48.2. I see the same as you with ktlint 0.49.0, but upgrading to that version is not possible for our project yet (we rely on a Gradle plugin that has not updated with 0.49.0 support). I now see that this is what you were referencing in https://github.com/pinterest/ktlint/issues/1933#issuecomment-1503647528; the formatting is applied correctly _despite the_ non-zero exit code.

Is there the possibility for a 0.48.3 release that includes this fix? If not we will keep an eye out for Spotless to support 0.49.0 (https://github.com/diffplug/spotless/issues/1696).

**Comment by paul-dingemans** [comment]
> Is there the possibility for a 0.48.3 release that includes this fix? If not we will keep an eye out for Spotless to support 0.49.0 ([diffplug/spotless#1696](https://github.com/diffplug/spotless/issues/1696)).

Sorry, but 0.48 will not be fixed for this.

**Comment by stkent** [comment]
> > Is there the possibility for a 0.48.3 release that includes this fix? If not we will keep an eye out for Spotless to support 0.49.0 ([diffplug/spotless#1696](https://github.com/diffplug/spotless/issues/1696)).
> 
> Sorry, but 0.48 will not be fixed for this.

No worries, thanks for confirming :) Appreciate the effort going into reaching 1.0!
